File: worker/src/features/room/lib/Broadcaster.ts
export type Lang = "en" | "ja" | "zh";
export const LANGS: Lang[] = ["en", "ja", "zh"];

export class Broadcaster {
  private hostWs: WebSocket | null = null;
  private guests = new Map<string, { ws: WebSocket; lang: Lang }>();
  private langGroups = new Map<Lang, Set<string>>(LANGS.map((l) => [l, new Set()]));

  setHost(ws: WebSocket) {
    this.hostWs = ws;
  }

  clearHost() {
    this.hostWs = null;
  }

  get hasHost(): boolean {
    return this.hostWs !== null;
  }

  addGuest(id: string, ws: WebSocket, lang: Lang) {
    this.guests.set(id, { ws, lang });
    this.langGroups.get(lang)!.add(id);
  }

  removeGuest(id: string) {
    const guest = this.guests.get(id);
    if (!guest) return;
    this.langGroups.get(guest.lang)?.delete(id);
    this.guests.delete(id);
  }

  activeLangs(): Lang[] {
    return LANGS.filter((lang) => (this.langGroups.get(lang)?.size ?? 0) > 0);
  }

  sendToHost(data: string) {
    if (this.hostWs) this.send(this.hostWs, data);
  }

  sendToLang(lang: Lang, data: string | Uint8Array) {
    for (const id of this.langGroups.get(lang) ?? []) {
      const guest = this.guests.get(id);
      if (guest) this.send(guest.ws, data);
    }
  }

  sendToAllGuests(data: string) {
    for (const [, guest] of this.guests) this.send(guest.ws, data);
  }

  sendToEveryone(data: string) {
    this.sendToHost(data);
    this.sendToAllGuests(data);
  }

  pushGuestCount() {
    const counts: Record<string, number> = {};
    for (const [lang, ids] of this.langGroups) counts[lang] = ids.size;
    this.sendToHost(JSON.stringify({ type: "room:guest_count", counts }));
  }

  private send(ws: WebSocket, data: string | Uint8Array) {
    try { ws.send(data); } catch {}
  }
}

File: worker-rs/src/room/broadcaster.rs
use std::collections::{HashMap, HashSet};
use worker::WebSocket;
use crate::types::{Lang, GuestCounts, ServerMsg};

struct Guest {
    ws: WebSocket,
    lang: Lang,
}

pub struct Broadcaster {
    host_ws: Option<WebSocket>,
    guests: HashMap<String, Guest>,
    lang_groups: HashMap<Lang, HashSet<String>>,
}

impl Broadcaster {
    pub fn new() -> Self {
        let mut lang_groups = HashMap::new();
        for lang in Lang::ALL {
            lang_groups.insert(lang, HashSet::new());
        }

        Self {
            host_ws: None,
            guests: HashMap::new(),
            lang_groups,
        }
    }

    // --- host ---

    pub fn has_host(&self) -> bool {
        self.host_ws.is_some()
    }

    pub fn set_host(&mut self, ws: WebSocket) {
        self.host_ws = Some(ws);
    }

    pub fn clear_host(&mut self) {
        self.host_ws = None;
    }

    // --- guests ---

    pub fn add_guest(&mut self, id: String, ws: WebSocket, lang: Lang) {
        self.lang_groups.entry(lang).or_default().insert(id.clone());
        self.guests.insert(id, Guest { ws, lang });
    }

    pub fn remove_guest(&mut self, id: &str) {
        if let Some(guest) = self.guests.remove(id) {
            if let Some(group) = self.lang_groups.get_mut(&guest.lang) {
                group.remove(id);
            }
        }
    }

    pub fn active_langs(&self) -> Vec<Lang> {
        Lang::ALL
            .iter()
            .copied()
            .filter(|lang| {
                self.lang_groups
                    .get(lang)
                    .map_or(false, |group| !group.is_empty())
            })
            .collect()
    }

    // --- send ---

    pub fn send_to_host(&self, data: &str) {
        if let Some(ws) = &self.host_ws {
            let _ = ws.send_with_str(data);
        }
    }

    pub fn send_to_lang_str(&self, lang: Lang, data: &str) {
        if let Some(ids) = self.lang_groups.get(&lang) {
            for id in ids {
                if let Some(guest) = self.guests.get(id) {
                    let _ = guest.ws.send_with_str(data);
                }
            }
        }
    }

    pub fn send_to_lang_bytes(&self, lang: Lang, data: &[u8]) {
        if let Some(ids) = self.lang_groups.get(&lang) {
            for id in ids {
                if let Some(guest) = self.guests.get(id) {
                    let _ = guest.ws.send_with_bytes(data);
                }
            }
        }
    }

    pub fn send_to_all_guests(&self, data: &str) {
        for guest in self.guests.values() {
            let _ = guest.ws.send_with_str(data);
        }
    }

    pub fn send_to_everyone(&self, data: &str) {
        self.send_to_host(data);
        self.send_to_all_guests(data);
    }

    pub fn push_guest_count(&self) {
        let counts = GuestCounts {
            en: self.lang_groups.get(&Lang::En).map_or(0, |g| g.len()),
            ja: self.lang_groups.get(&Lang::Ja).map_or(0, |g| g.len()),
            zh: self.lang_groups.get(&Lang::Zh).map_or(0, |g| g.len()),
        };

        let msg = serde_json::to_string(&ServerMsg::GuestCount { counts }).unwrap_or_default();
        self.send_to_host(&msg);
    }
}
